### ISSUE TITLE
trackers not blocked

### ISSUE THREAD

**Issue by fusionneur** [MuntashirAkon/AppManager#1071]
android 11
app freshly updated to version 8

blocking trackers is not working

tried clicking on trackers, then block or going to activities/services/receovers/providers then 3 dots menu block/unblock trackers. also toggling a tracker off does not apply and i get message "rules are not applied"

app was closed (force close was greyed out)

the trackers toggles remain on

i got similar behavior in the past for other apps, but sometimes i could get it work by force closing the app. all other apps can have the trackers properly blocked

**Comment by MuntashirAkon** [comment]
You need to be more precise in reporting the issue. If you can reproduce the issue, you need to explain it step by step so that we too can reproduce it. Otherwise, there's no point in reporting the issue because it won't be possible to fix it unless we find the underlying reasons.

**Comment by MuntashirAkon** [comment]
This is enough for now. I'll seriously look into it before the next release.

**Comment by MuntashirAkon** [comment]
May be fixed by b5b0041cde589b26dfcd178a18ae6a7b740f6512

**Comment by MuntashirAkon** [comment]
> was it just a visual bug or the trackers actually not blocked?

It was a caching error (if the above fix indeed worked). When you open App Manager, it starts syncing with the system in the background. So, if you opened an app with blocked trackers before the syncing was finished, it would misreport the blocked components. The issue is fixed by retrieving those components directly instead of relying on cache.

**Comment by shuvashish76** [comment]
AM latest Debug 👇🏿
![Screenshot_20230919-211956](https://github.com/MuntashirAkon/AppManager/assets/48226312/a68b5225-61c7-4904-a24f-9edc6d698c1f)

AM 4.0.0-alpha02 👇🏿
![Screenshot_20230919-212009](https://github.com/MuntashirAkon/AppManager/assets/48226312/c49db7a9-886d-4f21-befe-ab002ddc1aa0)

---
AM Debug list 👇🏿
![Screenshot_20230919-212401](https://github.com/MuntashirAkon/AppManager/assets/48226312/393abb4f-0dd4-451b-aee9-2f96cd75fa01)

AM 4.0.0-alpha02 list 👇🏿
![Screenshot_20230919-212611](https://github.com/MuntashirAkon/AppManager/assets/48226312/6a7763b6-fd0a-4fd5-82d4-e066ecaa28f8)

Observation:
1. Most trackers are blocked but the chip shows wrong colour for AM alpha. Uninstall-reinstall (alpha) > same wrong orange colour.
2. Why it shows some trackers in white for AM alpha?
3. After blocking again for AM alpha it shows correct Dark cyan coloured chip & all trackers in same colour.

**Comment by MuntashirAkon** [comment]
In the App Details page, App Manager retrieves info regarding component blocking from two sources:
1. App Manager itself
2. Intent Firewall

App Manager does not use any Android system API to fetch those info components can be _disabled_ for many reasons, not just for _blocking_ them. Hence, if you reinstall App Manager, the first source will no longer exist. So, it can only retrieve information from Intent Firewall. I'm guessing those components from the logger Sentry didn't exist in the Intent Firewall directory. Therefore, when you reinstalled the app, it could no longer mark them as blocked (though may be disabled). When you blocked them again via App Manager, it could again recognise them as blocked.

However, if they did exist in the directory (in <package-name>.xml) file, then this is a wrong behaviour and should be reported with steps to reproduce them.

**Comment by shuvashish76** [comment]
> If you reinstall App Manager, the first source will no longer exist.

Not related but similar issues #590 #591 ‽

> I'm guessing those components from the logger Sentry didn't exist in the Intent Firewall directory. Therefore, when you reinstalled the app, it could no longer mark them as blocked (though may be disabled).

Yes checked `/data/system/ifw` directory for that app, no sentry tracker entry in .xml file.

To make the discussion clear, we've 2 issues here.
1. ~Chips showing incorrect colour~ (Checked few other apps > shows dark cyan iff all trackers are blocked ✔️; while shows ~incorrect~ orange colour if some are not blocked ✔️). Now I realise that it shows correct info, this is not an issue for discussion.
2. Some trackers are not blocked when checked from the list (see below)

List of other trackers with no entry in the .xml file. (from my small list of installed apps)
<details><summary>list</summary>
<p>

```
io.invertase.firebase.crashlytics.ReactNativeFirebaseCrashlyticsInitProvider
com.google.mlkit.common.internal.MlKitInitProvider
io.sentry.android.core.SentryInitProvider
io.sentry.android.core.SentryPerformanceProvider
org.acra.attachment.AcraContentProvider
mwtins.de.appplant.cordova.emailcomposer.Provider
```

</p>
</details> 

One pattern I notice is AM always miss **all** trackers from 'Providers' component section.

---

Here is how it looks on both AM1 (operation: block trackers), AM2 (only install & see tracker status), reproducible by checking these foss apps e.g
<details><summary>Apps</summary>
<p>

- AnkiDroid (`org.acra.attachment.AcraContentProvider`)
- NewPipe (`org.acra.attachment.AcraContentProvider`)
- SchildiChat (`io.sentry.android.core.SentryInitProvider`, `io.sentry.android.core.SentryPerformanceProvider`)

</p>
</details> 

<details><summary>Screenshots</summary>
<p>

AM1
![Screenshot_20230920-052533](https://github.com/MuntashirAkon/AppManager/assets/48226312/6b365f4a-77c4-4f6f-b37e-f66fea951336)

AM2
![Screenshot_20230920-052603](https://github.com/MuntashirAkon/AppManager/assets/48226312/896fbf6f-8ec8-4460-a0d8-2874e9d18d4c)

</p>
</details> 

Different tracker status makes confusing if trackers are really blocked.

**Comment by MuntashirAkon** [comment]
> Not related but similar issues #590 #591 ‽

Similar to #590, yes. But that uninstalling App Manager destroys all information stored in App Manager is cannot be considered as a bug. It's the general behaviour like any other app you typically install. (I think many people has been misled by how blocking worked in other popular tools like MAT which is) blocking rules present in App Manager is NOT (it's ironical that Markdown does not support underlines) the same as blocking rules in the system because it's impossible to sync between them (See also, [FAQ](https://muntashirakon.github.io/AppManager/en/#sec:faq:app-components)).

> Different tracker status makes confusing if trackers are really blocked.

The colour code specifies whether a component is disabled. In both cases, the components are disabled. But one could recognise it as blocked (probably because you've blocked them in that app) and the other can't (you didn't block them via that app). App Manager and AM Debug are independent apps, and they do not communicate between each other. Because it's expected that people would use only one of those apps (preferably, the release version if they're not a tester), not both.

> One pattern I notice is AM always miss **all** trackers from 'Providers' component section.

This seems to be the only pattern I can see here. If you can confirm it by explicitly blocking them via IFW (by long clicking on the switch), then I can look into the issue. It might just be a compatibility issue as **Watt** used to mark providers as activities.

**Comment by shuvashish76** [comment]
>  If you can confirm it by explicitly blocking them via IFW (by long clicking on the switch), then I can look into the issue.

Checked (+reverse check) for 'Providers' with only IFW blocking same issue exist for AM1 & AM2. No entry in .xml file.

**Comment by MuntashirAkon** [comment]
Fixed via ddd31bc04067c373d66f0530cb0998dbb18b19b6
